java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java

```diff
@@ -582,39 +587,111 @@ private void executeMethodInvocation(MethodInvocationTree mit) {
-        programState = programState.resetFieldValues(constraintManager);
-      }
+      // Enqueue additional exceptional paths corresponding to unchecked exceptions, for instance OutOfMemoryError
+      enqueueUncheckedExceptionalPaths();
+
+      // Enqueue happy paths
+      programState = handleSpecialMethods(programState.stackValue(resultValue), mit, methodSymbol);
       checkerDispatcher.executeCheckPostStatement(mit);
       clearStack(mit);
     }
   }
 
-  private void enqueueExceptionalPaths(ProgramState ps) {
-    node.programPoint.block.exceptions().forEach(b -> enqueue(new ExplodedGraph.ProgramPoint(b, 0), ps.clearStack(), !b.isCatchBlock()));
+  private ProgramState handleSpecialMethods(ProgramState ps, MethodInvocationTree mit, Symbol methodSymbol) {
+    if (isNonNullMethod(methodSymbol)) {
+      return ps.addConstraint(ps.peekValue(), ObjectConstraint.NOT_NULL);
+    } else if (OBJECT_WAIT_MATCHER.matches(mit)) {
+      return ps.resetFieldValues(constraintManager);
+    }
+    return ps;
+  }
+
+  private void enqueueThrownExceptionalPaths(Symbol symbol) {
+    if (symbol.isMethodSymbol()) {
+      ProgramState ps = programState.clearStack();
+      ((Symbol.MethodSymbol) symbol).thrownTypes().stream()
+        .map(constraintManager::createExceptionalSymbolicValue)
+        .forEach(exceptionSV -> enqueueExceptionalPaths(ps, exceptionSV));
+    }
+  }
+
+  private void enqueueUncheckedExceptionalPaths() {
+    enqueueExceptionalPaths(programState.clearStack(), constraintManager.createExceptionalSymbolicValue(null));
+  }
+
+  private void enqueueExceptionalPaths(ProgramState ps, SymbolicValue.ExceptionalSymbolicValue exceptionSV) {
+    Set<CFG.Block> exceptionBlocks = node.programPoint.block.exceptions();
+    List<CFG.Block> catchBlocks = exceptionBlocks.stream().filter(CFG.Block.IS_CATCH_BLOCK).collect(Collectors.toList());
+
+    // only consider the first match, as order of catch block is important
+    Optional<CFG.Block> firstMatchingCatchBlock = catchBlocks.stream()
+      .filter(b -> isValidCatchBlock(b, exceptionSV.exceptionType()))
+      .sorted((b1, b2) -> Integer.compare(b2.id(), b1.id()))
+      .findFirst();
+    if (firstMatchingCatchBlock.isPresent()) {
+      enqueue(new ExplodedGraph.ProgramPoint(firstMatchingCatchBlock.get(), 0), ps);
+      return;
+    }
+
+    // branch to any unchecked exception catch
+    catchBlocks.stream()
+      .filter(ExplodedGraphWalker::isCatchingUncheckedException)
+      .forEach(b -> enqueue(new ExplodedGraph.ProgramPoint(b, 0), ps));
+
+    // use other exceptional blocks, i.e. finally block and exit blocks
+    exceptionBlocks.stream()
+      .filter(CFG.Block.IS_CATCH_BLOCK.negate())
+      .forEach(b -> enqueue(new ExplodedGraph.ProgramPoint(b, 0), ps, true));
+
+    // store the exception as exit value in case of method exit in next block
+    ProgramState newPs = ps.clearStack().stackValue(exceptionSV);
+    newPs.storeExitValue();
```

Comment:
This does not handle properly the case where you have a return in finally (and where you have not a return in the finally in fact)